mobx/test/observable_list_test.dart

```diff
@@ -98,20 +99,72 @@ void main() {
       expect(count, equals(2));
     });
 
+    test('observe set length works', () {
+      final list = ObservableList.of([0]);
+
+      var changesCount = 0;
+      var index = -1;
+      var addedValues = <int>[];
+      var removedValues = <int>[];
+
+      list.observe((change) {
+          changesCount += change.rangeChanges?.length ?? 0;
+          if (changesCount > 0) {
+            index = change.rangeChanges.keys.first;
+            addedValues = change.rangeChanges[index].newValues;
+            removedValues = change.rangeChanges[index].oldValues;
+          }
+        });
+
+      list.length = 5;
+      expect(changesCount, equals(1));
+      expect(index, equals(1));
+      expect(addedValues, equals(List.filled(4, null)));
+      expect(removedValues, equals(null));
+
+      list.length = 3;
+      expect(changesCount, equals(2));
+      expect(index, equals(3));
+      expect(addedValues, equals(null));
+      expect(removedValues, equals(List.filled(2, null)));
+
+      print('final size ${list.length}');
+    });
+
     test('observe insert item works', () {
       final list = ObservableList.of([1]);
 
+      var changesCount = -1;
       var index = -1;
       var addedValues = <int>[];
       var removedValues = <int>[];
 
       list
         ..observe((change) {
-          index = change.index;
-          addedValues = change.added;
-          removedValues = change.removed;
+          if (change.rangeChanges != null) {
+            changesCount = change.rangeChanges.length;
+            if (changesCount > 0) {
+              index = change.rangeChanges.keys.first;
+              addedValues = change.rangeChanges[index].newValues;
+              removedValues = change.rangeChanges[index].oldValues;
+            }
+          } else if (change.elementChanges != null) {
+            changesCount = change.elementChanges.length;
+            if (changesCount > 0) {
+              index = change.elementChanges.keys.first;
+              final firstChange = change.elementChanges[index];
+              addedValues = [OperationType.add, OperationType.update].contains(firstChange.type)
+                ? [firstChange.newValue]
+                : null;
+              removedValues = [OperationType.remove, OperationType.update].contains(firstChange.type)
+                ? [firstChange.oldValue]
+                : null;
+            }
+          }
```

Comment:
This is one of the example of the excessive conditional logic inside tests. The befor - after checks should be simpler IMHO. Let's have an explicitly created before / after states for comparison